### PR TITLE
Extract SetComponentObject into an extension method

### PR DESCRIPTION
**Pull request by firtina-improbable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
This makes it consistent with `GetComponentObject` extension method too.

Also adds some exceptions to help resolve problems if things change

#### Tests
Tested with local deployments and in editor

#### Documentation
Internal only

#### Primary reviewers
@zeroZshadow @jamiebrynes7

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+        static EntityManagerExtensions()
+        {
+            var setComponentObjectMethodInfo =
+                typeof(EntityManager).GetMethod("SetComponentObject", BindingFlags.Instance | BindingFlags.NonPublic,
+                    null,
+                    new Type[] { typeof(Entity), typeof(ComponentType), typeof(object) },
+                    new ParameterModifier[] { });
+
+            ValidateParameters(setComponentObjectMethodInfo);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: This should be wrapped in an `if` `endif` I think - as the `Conditional` attribute will exclude this method from compilation

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: AFAIK it turns the method call into a noop

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: That's surprising - but I've never used `Conditional` before so I trust you 😄 <sup>Pls no betray me</sup>

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            if (parameters.Length == 3 &&
+                parameters[0].ParameterType == typeof(Entity) &&
+                parameters[1].ParameterType == typeof(ComponentType) &&
+                parameters[2].ParameterType == typeof(object))
+            {
+                // All good!
+                return;
+            }
+
+            throw new MissingMethodException(errorMessage);
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: could we have a different error message here and include the exact signature of the new setComponentObject method

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs (hunk 1 above)
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: Can confirm what firtina says here.
This is why `Conditional` is used for loads of profiling related functions ;)

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            const string errorMessage =
+                "Could not find the EntityManager.SetComponentObject(Entity, ComponentType, object) method.\n" +
+                "Please ensure you are using a compatible version of the Unity.Entities package.";
+
+            if (setComponentObjectMethodInfo == null)
+            {
+                throw new MissingMethodException(errorMessage);
+            }
+
+            var parameters = setComponentObjectMethodInfo.GetParameters();
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: I'm not sure how this validates anything.
As `setComponentObjectMethodInfo`'s type declaration already ensures it has these parameters, or is null.

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                // All good!
+                return;
+            }
+
+            throw new MissingMethodException(errorMessage);
+        }
+
+        public static void SetComponentObject<T>(this EntityManager entityManager, Entity entity, T component)
+        {
+            SetComponentObjectAction(entityManager, entity, typeof(T), component);
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: Please use `ComponentType.Create<T>()` here.

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs (hunk 3 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: Good point, I am blind

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs (hunk 4 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: Roger

File: workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                typeof(EntityManager).GetMethod("SetComponentObject", BindingFlags.Instance | BindingFlags.NonPublic,
+                    null,
+                    new Type[] { typeof(Entity), typeof(ComponentType), typeof(object) },
+                    new ParameterModifier[] { });
+
+            if (setComponentObjectMethodInfo == null)
+            {
+                throw new MissingMethodException(
+                    "Could not find the EntityManager.SetComponentObject(Entity, ComponentType, object) method.\n" +
+                    "Please ensure you are using a compatible version of the Unity.Entities package.");
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: What version would that be? Help the users out :)

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: Currently it's a moving target. I'll add a ticket to specify version once we're in the polish phase.

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/Utility/EntityManagerExtensions.cs]: UTY-797

**Pull request merged**